From f484a46574eefc7124925ebb040a9662d638302b Mon Sep 17 00:00:00 2001
From: Fancy Fang <chen.fang@nxp.com>
Date: Wed, 13 Jul 2016 09:40:54 +0800
Subject: [PATCH 1112/5242] MLK-12983 dma: pxp: add sanity check in
 pxp_set_scaling()

commit  73877d9b4ede0f9b3fec2b84d85255e2b0a74cd2 from
https://source.codeaurora.org/external/imx/linux-imx.git

Add the sanity check for drect's width and height in
pxp_set_scaling() to avoid div0 exception.

Signed-off-by: Fancy Fang <chen.fang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/dma/pxp/pxp_dma_v2.c |    6 ++++++
 drivers/dma/pxp/pxp_dma_v3.c |    6 ++++++
 2 files changed, 12 insertions(+)

diff --git a/drivers/dma/pxp/pxp_dma_v2.c b/drivers/dma/pxp/pxp_dma_v2.c
index 0128881..0b12a10 100644
--- a/drivers/dma/pxp/pxp_dma_v2.c
+++ b/drivers/dma/pxp/pxp_dma_v2.c
@@ -659,6 +659,12 @@ static int pxp_set_scaling(struct pxps *pxp)
 	struct pxp_layer_param *out_params = &pxp_conf->out_param;
 
 	proc_data->scaling = 1;
+
+	if (!proc_data->drect.width || !proc_data->drect.height) {
+		pr_err("Invalid drect width and height passed in\n");
+		return -EINVAL;
+	}
+
 	decx = proc_data->srect.width / proc_data->drect.width;
 	decy = proc_data->srect.height / proc_data->drect.height;
 	if (decx > 1) {
diff --git a/drivers/dma/pxp/pxp_dma_v3.c b/drivers/dma/pxp/pxp_dma_v3.c
index aea7541..8d05f50 100644
--- a/drivers/dma/pxp/pxp_dma_v3.c
+++ b/drivers/dma/pxp/pxp_dma_v3.c
@@ -851,6 +851,12 @@ static int pxp_set_scaling(struct pxps *pxp)
 	struct pxp_layer_param *out_params = &pxp_conf->out_param;
 
 	proc_data->scaling = 1;
+
+	if (!proc_data->drect.width || !proc_data->drect.height) {
+		pr_err("Invalid drect width and height passed in\n");
+		return -EINVAL;
+	}
+
 	decx = proc_data->srect.width / proc_data->drect.width;
 	decy = proc_data->srect.height / proc_data->drect.height;
 	if (decx > 1) {
-- 
1.7.9.5

